File: windows-native-launcher/WakeWordClassifier.cs
using System;
using System.Linq;
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;

namespace Mana.NativeLauncher;

// Acoustic pre-filter for wake-word detection (#342): runs on every
// VAD-closed segment before it's sent to Whisper, so segments that don't
// acoustically contain "Mana" never pay the transcription cost. This is a
// soft optimization, not the wake-up decision itself -- WakeWordMatcher's
// existing text-based check on the (now-conditionally-reached) transcript
// still has final say, per the #342 integration design: a false acoustic
// trigger only ever wastes one Whisper call, it can never cause a false
// wake-up on its own.
//
// Chains three ONNX models exactly as openWakeWord's own Python
// `AudioFeatures`/training code does -- ported from, not reimplemented
// independently of, tts-service/venv/Lib/site-packages/openwakeword/
// utils.py's `_get_melspectrogram`/`_get_embeddings`, and verified against
// the real ONNX models' actual input/output shapes at each stage (not
// assumed from the Python source alone -- see tools/wakeword-training/'s
// own PRs for how this was confirmed):
//
//   raw 16-bit PCM audio (1, samples)
//     -> melspectrogram.onnx -> (1, 1, time, 32), squeezed to (time, 32),
//        transformed by x/10+2 to match the original TF implementation
//     -> a sliding window of 76 mel-frames, stepped by 8, over the time
//        axis -- each full window becomes one (76, 32, 1) input
//     -> embedding_model.onnx, batched over all windows -> (windows, 1, 1, 96),
//        squeezed to (windows, 96)
//     -> mana.onnx (ours, trained in tools/wakeword-training/) -> (1, 1)
//        sigmoid score, when windows == EmbeddingFrames (16)
//
// melspectrogram.onnx and embedding_model.onnx are openWakeWord's own
// third-party pretrained models, fetched at build time (same convention
// as silero_vad.onnx -- see the .csproj). mana.onnx is ours, committed
// directly since nothing else hosts it.
internal sealed class WakeWordClassifier : IDisposable
{
    internal const int SampleRate = 16000;

    // The fixed embedding-frame window mana.onnx was trained on --
    // confirmed empirically in tools/wakeword-training/train_model.py by
    // calling AudioFeatures.get_embedding_shape() at increasing window
    // lengths until it matched the training data's own shape, not assumed.
    internal const int EmbeddingFrames = 16;

    // 2.0 seconds at 16kHz -- the audio length that produces exactly
    // EmbeddingFrames embedding windows (197 mel-frames -> 16 windows of
    // 76 stepped by 8), confirmed against the real ONNX models.
    internal const int WindowSamples = SampleRate * 2;

    // See tools/wakeword-training/README.md's measured false-positive-
    // rate-by-threshold table: 7.2/hour at 0.5, 1.0/hour at 0.99. 0.9 is a
    // reasonable middle ground for a v1 model trained on a modest amount
    // of positive data -- configurable via ManaSettingsStore, not a fixed
    // constant callers are stuck with.
    internal const float DefaultThreshold = 0.9f;

    private const int MelWindowSize = 76;
    private const int MelStepSize = 8;
    private const int MelBins = 32;
    private const int EmbeddingDim = 96;

    private readonly InferenceSession melspecSession;
    private readonly InferenceSession embeddingSession;
    private readonly InferenceSession classifierSession;
    private readonly float threshold;

    public WakeWordClassifier(
        string melspecModelPath,
        string embeddingModelPath,
        string classifierModelPath,
        float threshold = DefaultThreshold)
    {
        melspecSession = new InferenceSession(melspecModelPath);
        embeddingSession = new InferenceSession(embeddingModelPath);
        classifierSession = new InferenceSession(classifierModelPath);
        this.threshold = threshold;
    }

    // samples: 16-bit PCM at 16kHz, any length. Right-aligned/padded to
    // exactly WindowSamples to match how the training clips were aligned
    // (the wake phrase ends near the window's end, with a small start
    // jitter -- see train_model.py's `starts` calculation) -- mana.onnx
    // only ever saw fixed 2.0s windows during training, so a real VAD
    // segment of a different length needs to land in the same place a
    // training example would have.
    public float Score(short[] samples)
    {
        var windowed = AlignToWindow(samples);
        var melspec = Melspectrogram(windowed);
        var embeddings = Embed(melspec);

        if (embeddings.GetLength(0) != EmbeddingFrames)
        {
            // A 2.0s input should always produce exactly 16 windows against
            // these specific models -- if it doesn't, something upstream
            // changed shape in a way this port didn't anticipate. Fail
            // loudly rather than feed a wrong-shaped tensor into the
            // classifier and get a meaningless score back.
            throw new InvalidOperationException(
                $"Expected {EmbeddingFrames} embedding windows from a {WindowSamples}-sample input, got {embeddings.GetLength(0)}.");
        }

        return Classify(embeddings);
    }

    public bool MayContainWakeWord(short[] samples) => Score(samples) >= threshold;

    private static short[] AlignToWindow(short[] samples)
    {
        if (samples.Length == WindowSamples)
        {
            return samples;
        }

        var aligned = new short[WindowSamples];
        if (samples.Length > WindowSamples)
        {
            // Longest final WindowSamples -- matches training's
            // end-aligned convention when the real segment ran longer
            // than the model's training window.
            Array.Copy(samples, samples.Length - WindowSamples, aligned, 0, WindowSamples);
        }
        else
        {
            // Left-pad with silence so the (shorter) real segment still
            // ends at the window's end, same alignment as training.
            Array.Copy(samples, 0, aligned, WindowSamples - samples.Length, samples.Length);
        }

        return aligned;
    }

    // Raw int16 samples cast to float32, NOT normalized to [-1,1] -- this
    // model's own training/inference convention (confirmed against
    // openwakeword.utils.AudioFeatures._get_melspectrogram, which raises
    // if given anything other than int16-range data).
    private float[,] Melspectrogram(short[] samples)
    {
        var floatSamples = new float[samples.Length];
        for (int i = 0; i < samples.Length; i++)
        {
            floatSamples[i] = samples[i];
        }

        var inputTensor = new DenseTensor<float>(floatSamples, new[] { 1, samples.Length });
        var inputs = new[] { NamedOnnxValue.CreateFromTensor("input", inputTensor) };

        using var results = melspecSession.Run(inputs);
        var output = results.First().AsTensor<float>();

        // Real shape confirmed as (1, 1, time, 32) -- the two leading
        // size-1 dims contribute nothing to a flat-array stride, so a
        // flat[t * MelBins + b] index is equivalent to squeezing them
        // away first, without needing to actually reshape the array.
        var dims = output.Dimensions.ToArray();
        int frames = dims[dims.Length - 2];
        var flat = output.ToArray();

        var mel = new float[frames, MelBins];
        for (int t = 0; t < frames; t++)
        {
            for (int b = 0; b < MelBins; b++)
            {
                // The fixed transform openWakeWord applies to match the
                // original TensorFlow implementation's scale.
                mel[t, b] = flat[(t * MelBins) + b] / 10f + 2f;
            }
        }

        return mel;
    }

    // Slides a MelWindowSize-frame window over the mel spectrogram's time
    // axis, stepped by MelStepSize, dropping any short trailing window --
    // matches openwakeword.utils.AudioFeatures._get_embeddings exactly.
    private float[,] Embed(float[,] melspec)
    {
        int totalFrames = melspec.GetLength(0);
        var windowStarts = new System.Collections.Generic.List<int>();
        for (int i = 0; i + MelWindowSize <= totalFrames; i += MelStepSize)
        {
            windowStarts.Add(i);
        }

        int windowCount = windowStarts.Count;
        var batchData = new float[windowCount * MelWindowSize * MelBins];
        for (int w = 0; w < windowCount; w++)
        {
            int start = windowStarts[w];
            for (int t = 0; t < MelWindowSize; t++)
            {
                for (int b = 0; b < MelBins; b++)
                {
                    batchData[(w * MelWindowSize * MelBins) + (t * MelBins) + b] = melspec[start + t, b];
                }
            }
        }

        var inputTensor = new DenseTensor<float>(batchData, new[] { windowCount, MelWindowSize, MelBins, 1 });
        var inputs = new[] { NamedOnnxValue.CreateFromTensor("input_1", inputTensor) };

        using var results = embeddingSession.Run(inputs);
        var output = results.First().AsTensor<float>();
        var flat = output.ToArray();

        // Real shape confirmed as (windows, 1, 1, 96) -- same
        // stride-through-size-1-dims reasoning as Melspectrogram above.
        var embeddings = new float[windowCount, EmbeddingDim];
        for (int w = 0; w < windowCount; w++)
        {
            for (int d = 0; d < EmbeddingDim; d++)
            {
                embeddings[w, d] = flat[(w * EmbeddingDim) + d];
            }
        }

        return embeddings;
    }

    private float Classify(float[,] embeddings)
    {
        int frames = embeddings.GetLength(0);
        var data = new float[frames * EmbeddingDim];
        for (int t = 0; t < frames; t++)
        {
            for (int d = 0; d < EmbeddingDim; d++)
            {
                data[(t * EmbeddingDim) + d] = embeddings[t, d];
            }
        }

        var inputTensor = new DenseTensor<float>(data, new[] { 1, frames, EmbeddingDim });
        var inputs = new[] { NamedOnnxValue.CreateFromTensor("input", inputTensor) };

        using var results = classifierSession.Run(inputs);
        return results.First().AsTensor<float>().ToArray()[0];
    }

    public void Dispose()
    {
        melspecSession.Dispose();
        embeddingSession.Dispose();
        classifierSession.Dispose();
    }
}

File: windows-launcher/renderer/wakeword-classifier.js
// Acoustic pre-filter for wake-word detection (#342): mirrors
// silero-vad.js's factory/injected-`ort` pattern, but chains three ONNX
// sessions instead of one, exactly as openWakeWord's own Python
// AudioFeatures class does. Ported from, and verified against, the real
// models' actual input/output shapes (confirmed via onnxruntime's own
// introspection plus real numeric Python runs before writing a line of
// this file -- see windows-native-launcher/WakeWordClassifier.cs's
// header comment for the full verification this was ported from, not
// re-derived independently):
//
//   raw 16-bit PCM audio (1, samples)
//     -> melspectrogram.onnx -> (1, 1, time, 32), squeezed to (time, 32),
//        transformed by x/10+2 to match the original TF implementation
//     -> a sliding window of 76 mel-frames, stepped by 8, over the time
//        axis -- each full window becomes one (76, 32, 1) input
//     -> embedding_model.onnx, batched over all windows -> (windows, 1, 1, 96),
//        squeezed to (windows, 96)
//     -> mana.onnx (ours, trained in tools/wakeword-training/) -> (1, 1)
//        sigmoid score, when windows === EMBEDDING_FRAMES (16)
//
// melspectrogram.onnx/embedding_model.onnx are openWakeWord's own
// third-party pretrained models (fetched at build/install time, same
// convention as silero_vad.onnx -- see scripts/fetch-silero-vad.js's
// sibling for these). mana.onnx is ours, committed directly since nothing
// else hosts it.
const SAMPLE_RATE = 16000;

// The fixed embedding-frame window mana.onnx was trained on -- confirmed
// empirically in tools/wakeword-training/train_model.py, not assumed.
const EMBEDDING_FRAMES = 16;

// 2.0 seconds at 16kHz -- the audio length that produces exactly
// EMBEDDING_FRAMES embedding windows against the real ONNX models.
const WINDOW_SAMPLES = SAMPLE_RATE * 2;

const MEL_WINDOW_SIZE = 76;
const MEL_STEP_SIZE = 8;
const MEL_BINS = 32;
const EMBEDDING_DIM = 96;

// See tools/wakeword-training/README.md's measured false-positive-rate-
// by-threshold table: 7.2/hour at 0.5, 1.0/hour at 0.99. 0.9 is a
// reasonable middle ground for a v1 model.
const DEFAULT_THRESHOLD = 0.9;

function createWakeWordClassifier({
  ort,
  melspecModelUrl,
  embeddingModelUrl,
  classifierModelUrl,
  threshold = DEFAULT_THRESHOLD,
} = {}) {
  if (!ort) {
    throw new Error("createWakeWordClassifier requires an onnxruntime-web `ort` instance");
  }
  if (!melspecModelUrl || !embeddingModelUrl || !classifierModelUrl) {
    throw new Error(
      "createWakeWordClassifier requires melspecModelUrl, embeddingModelUrl, and classifierModelUrl",
    );
  }

  let sessionsPromise = null;

  function load() {
    if (!sessionsPromise) {
      sessionsPromise = Promise.all([
        ort.InferenceSession.create(melspecModelUrl),
        ort.InferenceSession.create(embeddingModelUrl),
        ort.InferenceSession.create(classifierModelUrl),
      ]);
    }
    return sessionsPromise;
  }

  // samples: Int16Array at 16kHz, any length. Right-aligned/padded to
  // exactly WINDOW_SAMPLES to match how the training clips were aligned
  // (the wake phrase ends near the window's end, with a small start
  // jitter -- see train_model.py's `starts` calculation) -- mana.onnx
  // only ever saw fixed 2.0s windows during training.
  function alignToWindow(samples) {
    if (samples.length === WINDOW_SAMPLES) {
      return samples;
    }
    const aligned = new Int16Array(WINDOW_SAMPLES);
    if (samples.length > WINDOW_SAMPLES) {
      // Rightmost WINDOW_SAMPLES -- matches training's end-aligned
      // convention when the real segment ran longer than the model's
      // training window.
      aligned.set(samples.subarray(samples.length - WINDOW_SAMPLES));
    } else {
      // Left-pad with silence so the (shorter) real segment still ends
      // at the window's end, same alignment as training.
      aligned.set(samples, WINDOW_SAMPLES - samples.length);
    }
    return aligned;
  }

  // Raw int16 samples cast to float32, NOT normalized to [-1,1] -- this
  // model's own training/inference convention (confirmed against
  // openwakeword.utils.AudioFeatures._get_melspectrogram).
  async function melspectrogram(session, samples) {
    const floatSamples = Float32Array.from(samples);
    const feeds = { input: new ort.Tensor("float32", floatSamples, [1, floatSamples.length]) };
    const results = await session.run(feeds);
    const output = results.output;

    // Real shape confirmed as (1, 1, time, 32) -- the two leading size-1
    // dims contribute nothing to a flat-array stride, so a
    // flat[t * MEL_BINS + b] index is equivalent to squeezing them away
    // first, without needing to actually reshape the array.
    const dims = output.dims;
    const frames = dims[dims.length - 2];
    const flat = output.data;
    const mel = new Float32Array(frames * MEL_BINS);
    for (let t = 0; t < frames; t += 1) {
      for (let b = 0; b < MEL_BINS; b += 1) {
        // The fixed transform openWakeWord applies to match the
        // original TensorFlow implementation's scale.
        mel[t * MEL_BINS + b] = flat[t * MEL_BINS + b] / 10 + 2;
      }
    }
    return { mel, frames };
  }

  // Slides a MEL_WINDOW_SIZE-frame window over the mel spectrogram's time
  // axis, stepped by MEL_STEP_SIZE, dropping any short trailing window --
  // matches openwakeword.utils.AudioFeatures._get_embeddings exactly.
  async function embed(session, mel, frames) {
    const windowStarts = [];
    for (let i = 0; i + MEL_WINDOW_SIZE <= frames; i += MEL_STEP_SIZE) {
      windowStarts.push(i);
    }

    const windowCount = windowStarts.length;
    const batchData = new Float32Array(windowCount * MEL_WINDOW_SIZE * MEL_BINS);
    for (let w = 0; w < windowCount; w += 1) {
      const start = windowStarts[w];
      for (let t = 0; t < MEL_WINDOW_SIZE; t += 1) {
        for (let b = 0; b < MEL_BINS; b += 1) {
          batchData[w * MEL_WINDOW_SIZE * MEL_BINS + t * MEL_BINS + b] = mel[(start + t) * MEL_BINS + b];
        }
      }
    }

    const feeds = {
      input_1: new ort.Tensor("float32", batchData, [windowCount, MEL_WINDOW_SIZE, MEL_BINS, 1]),
    };
    const results = await session.run(feeds);
    const output = results.conv2d_19;
    const flat = output.data;

    // Real shape confirmed as (windows, 1, 1, 96) -- same stride-through-
    // size-1-dims reasoning as melspectrogram() above.
    const embeddings = new Float32Array(windowCount * EMBEDDING_DIM);
    for (let w = 0; w < windowCount; w += 1) {
      for (let d = 0; d < EMBEDDING_DIM; d += 1) {
        embeddings[w * EMBEDDING_DIM + d] = flat[w * EMBEDDING_DIM + d];
      }
    }
    return { embeddings, windowCount };
  }

  async function classify(session, embeddings, frames) {
    const feeds = { input: new ort.Tensor("float32", embeddings, [1, frames, EMBEDDING_DIM]) };
    const results = await session.run(feeds);
    return results.output.data[0];
  }

  async function score(samples) {
    const [melspecSession, embeddingSession, classifierSession] = await load();
    const aligned = alignToWindow(samples);
    const { mel, frames } = await melspectrogram(melspecSession, aligned);
    const { embeddings, windowCount } = await embed(embeddingSession, mel, frames);

    if (windowCount !== EMBEDDING_FRAMES) {
      // A 2.0s input should always produce exactly 16 windows against
      // these specific models -- if it doesn't, something upstream
      // changed shape in a way this port didn't anticipate. Fail loudly
      // rather than feed a wrong-shaped tensor into the classifier and
      // get a meaningless score back.
      throw new Error(
        `Expected ${EMBEDDING_FRAMES} embedding windows from a ${WINDOW_SAMPLES}-sample input, got ${windowCount}.`,
      );
    }

    return classify(classifierSession, embeddings, windowCount);
  }

  async function mayContainWakeWord(samples) {
    return (await score(samples)) >= threshold;
  }

  return { load, score, mayContainWakeWord, alignToWindow };
}

module.exports = {
  createWakeWordClassifier,
  SAMPLE_RATE,
  EMBEDDING_FRAMES,
  WINDOW_SAMPLES,
  DEFAULT_THRESHOLD,
};
